arc_cache.go

```diff
@@ -28,28 +33,76 @@ func newARCCachedBS(ctx context.Context, bs Blockstore, lruSize int) (*arccache,
 	if err != nil {
 		return nil, err
 	}
-	c := &arccache{arc: arc, blockstore: bs}
+	arcLks := make(map[cid.Cid]*sync.Mutex)
+	c := &arccache{arc: arc, arcLks: arcLks, blockstore: bs}
 	c.hits = metrics.NewCtx(ctx, "arc.hits_total", "Number of ARC cache hits").Counter()
 	c.total = metrics.NewCtx(ctx, "arc_total", "Total number of ARC cache requests").Counter()
 
 	return c, nil
 }
 
-func (b *arccache) DeleteBlock(k cid.Cid) error {
-	if has, _, ok := b.hasCached(k); ok && !has {
-		return nil
+// if ok == false has is inconclusive, calling release is a noop
+// if ok == true then has response to question: is it contained and a lock will be held on content `k` until `release` is called.
+// it is the callers responsibility to call `release` when they are done writing/reading data.
+func (b *arccache) hasCachedSync(k cid.Cid) (has bool, size int, ok bool, release func()) {
+	// default return ops
+	has = false
+	size = -1
+	ok = false
+	release = func() {}
+
+	b.total.Inc()
+	if !k.Defined() {
+		log.Error("undefined cid in arccache")
+		// Return cache invalid so the call to blockstore happens
+		// in case of invalid key and correct error is created.
+		return
 	}
 
-	b.arc.Remove(k) // Invalidate cache before deleting.
-	err := b.blockstore.DeleteBlock(k)
-	if err == nil {
-		b.cacheHave(k, false)
+	h, ok := b.arc.Get(string(k.Hash()))
+	if ok {
+		b.hits.Inc()
+		switch h := h.(type) {
+		case cacheHave:
+			has = bool(h)
+			size = -1
+			ok = true
+		case cacheSize:
+			has = true
+			size = int(h)
+			ok = true
+		}
 	}
-	return err
+
+	// lock the map of cid->locks.
+	b.arcLksMu.Lock()
+	defer b.arcLksMu.Unlock()
+
+	// check if we have a lock for this content.
+	lk, hasLk := b.arcLks[k]
+	if has && hasLk {
+		// cache and lock hit.
+		lk.Lock()
```

Comment:
If another goroutine has already locked this CID, then locking here will wait to acquire the lock for the CID.  However, `b.arcLksMu` is not unlocked until this function exits.  This means that no other CID can be locked if there is more than one lock on the same CID.

Reply:
oof, thanks for catching this. I'll swap this out for an RWMutex in a subsequent commit.

Reply:
As an alternative to this, I've filed https://github.com/ipfs/go-ipfs-blockstore/pull/68 that uses a `sync.Map`.